[no-unused-vars] false-positive for class expressions

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to master, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->
 


```
$ git clone https://github.com/ocavue/typescript-eslint-issue-2831 /tmp/eslint-issue
$ cd /tmp/eslint-issue
$ yarn install 
$ yarn run lint 
```

**Expected Result**

No warning.

<!--
What did you expect to happen?
Please be specific here - list the exact lines and messages you expect.
-->

**Actual Result**



<!--
What actually happened?
Please be specific here - list the exact lines and messages that caused errors
-->


```
  2:11  warning  'C1' is defined but never used  @typescript-eslint/no-unused-vars
  4:11  warning  'C2' is defined but never used  @typescript-eslint/no-unused-vars
  6:11  warning  'C3' is defined but never used  @typescript-eslint/no-unused-vars
```

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

If I downgrade `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to version `4.8.2`, no warning occurs.
**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.0` |
| `@typescript-eslint/parser`        | `4.9.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `15.2.1` |

--- a/packages/eslint-plugin/src/util/collectUnusedVariables.ts
+++ b/packages/eslint-plugin/src/util/collectUnusedVariables.ts
@@ -153,6 +153,19 @@ class UnusedVarsVisitor<
     }
   }
 
+  private visitClass(
+    node: TSESTree.ClassDeclaration | TSESTree.ClassExpression,
+  ): void {
+    // skip a variable of class itself name in the class scope
+    const scope = this.getScope<TSESLint.Scope.Scopes.ClassScope>(node);
+    for (const variable of scope.variables) {
+      if (variable.identifiers[0] === scope.block.id) {
+        this.markVariableAsUsed(variable);
+        return;
+      }
+    }
+  }
+
   private visitFunction(
     node: TSESTree.FunctionDeclaration | TSESTree.FunctionExpression,
   ): void {
@@ -201,16 +214,9 @@ class UnusedVarsVisitor<
   //#region VISITORS
   // NOTE - This is a simple visitor - meaning it does not support selectors
 
-  protected ClassDeclaration(node: TSESTree.ClassDeclaration): void {
-    // skip a variable of class itself name in the class scope
-    const scope = this.getScope<TSESLint.Scope.Scopes.ClassScope>(node);
-    for (const variable of scope.variables) {
-      if (variable.identifiers[0] === scope.block.id) {
-        this.markVariableAsUsed(variable);
-        return;
-      }
-    }
-  }
+  protected ClassDeclaration = this.visitClass;
+
+  protected ClassExpression = this.visitClass;
 
   protected FunctionDeclaration = this.visitFunction;
 

--- a/packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts
@@ -960,6 +960,8 @@ interface _Foo {
       // ignored by pattern, even though it's only self-referenced
       options: [{ varsIgnorePattern: '^_' }],
     },
+    // https://github.com/ocavue/typescript-eslint-issue-2831/blob/master/index.ts
+    'export const classes = [class C1 {}, class C2 {}, class C3 {}];',
   ],
 
   invalid: [

